src/rust/engine/process_execution/src/speculate.rs

```diff
@@ -0,0 +1,218 @@
+use super::{CommandRunner, ExecuteProcessRequest, FallibleExecuteProcessResult};
+use boxfuture::{BoxFuture, Boxable};
+use futures::future::{err, ok, Future};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio_timer::Delay;
+use workunit_store::WorkUnitStore;
+
+#[derive(Clone)]
+pub struct SpeculatingCommandRunner {
+  primary: Arc<dyn CommandRunner>,
+  secondary: Arc<dyn CommandRunner>,
+  speculation_timeout: Duration,
+}
+
+impl SpeculatingCommandRunner {
+  pub fn new(
+    primary: Box<dyn CommandRunner>,
+    secondary: Box<dyn CommandRunner>,
+    speculation_timeout: Duration,
+  ) -> SpeculatingCommandRunner {
+    SpeculatingCommandRunner {
+      primary: primary.into(),
+      secondary: secondary.into(),
+      speculation_timeout: speculation_timeout,
+    }
+  }
+}
+
+impl CommandRunner for SpeculatingCommandRunner {
+  fn run(
+    &self,
+    req: ExecuteProcessRequest,
+    workunit_store: WorkUnitStore,
+  ) -> BoxFuture<FallibleExecuteProcessResult, String> {
+    let command_runner = self.clone();
+    let workunit_store_clone = workunit_store.clone();
+    let req_2 = req.clone();
+    let delay = Delay::new(Instant::now() + self.speculation_timeout);
+    self
+      .primary
+      .run(req, workunit_store)
+      .select(delay.then(move |_| command_runner.secondary.run(req_2, workunit_store_clone)))
+      .then(|raced_result| match raced_result {
+        Ok((successful_res, _outstanding_req)) => {
+          ok::<FallibleExecuteProcessResult, String>(successful_res).to_boxed()
+        }
+        Err((failed_res, _outstanding_req)) => {
+          err::<FallibleExecuteProcessResult, String>(failed_res).to_boxed()
+        }
+      })
+      .to_boxed()
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use crate::remote::tests::echo_foo_request;
+  use boxfuture::{BoxFuture, Boxable};
+  use bytes::Bytes;
+  use futures::future::Future;
+  use hashing::EMPTY_DIGEST;
+  use std::sync::{Arc, Mutex};
+  use std::time::{Duration, Instant};
+  use tokio;
+  use tokio_timer::Delay;
+  use workunit_store::WorkUnitStore;
+
+  use super::{
+    CommandRunner, ExecuteProcessRequest, FallibleExecuteProcessResult, SpeculatingCommandRunner,
+  };
+
+  #[test]
+  fn test_no_speculation() {
+    let (result, call_counter) = run_speculation_test(0, 0, 100, false, false);
+    assert_eq![1, *call_counter.lock().unwrap()];
```

Comment:
For the no speculation case, it may actually be nice to have two counters; a "started" counter and a "finished" counter, to show that we didn't even start the second attempt. But also understand if you don't think it's worthwhile :)